### PR TITLE
Bribes redirection

### PR DESCRIPTION
**Pull request by sajanrajdev**
### Changes introduced
- Removed hardcoded redirection of BADGER to BadgerTree
- Mapping of tokens to be redirected to their respective destinations and the redirection fees to be processed for each of these paths. This is maintainable by governance.
- Any token added to the list mentioned above will be redirected upon bribes claiming after its fee is processed, otherwise they will be sent to the BribesProcessor directly
- Events were included for the token redirections (to better inform of on-chain operations to partner DAOs and our treasury) and the redirection fees (to help track revenue streams).
- All of the above was tested including the upgrade storage integrity (additional missing tests were also introduced for the sweeping of rewards and some of the permissions)

### Notes
- As per discussions with @Tritium-VLK, it was decided to not introduce the redirection split feature as it is highly unlikely to be utilized and adding it meant a significant amount added of manual maintenance work for every bribes round as well as meaningful gas usage increase.
- The current setup allows for simple redirections of certain rewards to certain destinations and fee processing for this service. This should be enough given that HiddenHands will soon implement a whitelist to only allow our parterns to bribe with their token. While this list gets implemented, it is still unlikely for a random briber to bribe using another DAO's token and, if they do, it is highly unlikely that the amounts will be significant enough for it to justify the bi-weekly extra work + gas expenditure. It will be communicated and made clear enough that bribes set to be redirected will do so for the total amount obtained by graviAURA
- ETH can't be redirected (wouldn't make sense anyway)
- Redirection fee can be anything between 0 and 100%, do we want to hardcode a cap?

### Testing
All tests are passing:
image
image

### CODE REVIEW

File: contracts/MyStrategy.sol [hunk 1]
```diff
@@ -453,18 +497,39 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
...
+                redirectionFee,
+                block.number,
+                block.timestamp
+            );
+        }
+
+        // Send remaining to bribe recepient
+        // NOTE: Calculating instead of checking balance since there could have been an 
+        // existing balance on the contract beforehand
+        IERC20Upgradeable(token).safeTransfer(recepient, amount.sub(redirectionFee));
```
Comment (rayeaster) [inline contracts/MyStrategy.sol:518]: do we need a non-zero check before `safeTransfer`?

File: contracts/MyStrategy.sol [hunk 2]
```diff
@@ -453,18 +497,39 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
...
+            block.timestamp
+        );
     }
 
     function _sweepRewardToken(address token) internal {
         _onlyNotProtectedTokens(token);
 
         uint256 toSend = IERC20Upgradeable(token).balanceOf(address(this));
-        _handleRewardTransfer(token, toSend);
+        _sendTokenToBribesProcessor(token, toSend);
```
Comment (shuklaayush) [inline contracts/MyStrategy.sol:532]: Why this change? If someone else claims the bribes for us, then there would be no way to send them to the redirection contracts. We'd only be able to send it to the bribes processor

File: contracts/MyStrategy.sol [hunk 3]
```diff
@@ -349,15 +392,16 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
                 if (difference > 0) {
                     IWeth(address(WETH)).deposit{value: difference}();
                     nonZeroDiff = true;
-                    _handleRewardTransfer(address(WETH), difference);
+                    _handleRewardTransfer(address(WETH), address(0), difference);
```
Comment (shuklaayush) [inline contracts/MyStrategy.sol:395]: Is eth always going to be forwarded to the bribe processor?

File: contracts/MyStrategy.sol [hunk 4]
```diff
@@ -100,26 +122,47 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
...
+    function setRedirectionToken(address token, address recepient, uint256 redirectionFee) external {
+        _onlyGovernance();
+        require(token != address(0), "Invalid token address");
+        require(recepient != address(0), "Invalid recepient address");
+        require(redirectionFee <= MAX_BPS, "Invalid redirection fee");
+
+        // Sets redirection path for a given token
+        bribesRedirectionPaths[token] = recepient;
+        // Sets redirection fees for a given recepient
+        redirectionFees[recepient] = redirectionFee;
```
Comment (shuklaayush) [inline contracts/MyStrategy.sol:161]: Why's the fee based on recipient and not tokens? Also, might be better to have a separate function for setting redirection fees if they are based only on recipients

File: contracts/MyStrategy.sol [hunk 5]
```diff
@@ -56,13 +59,32 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
...
+        uint256 indexed blockNumber,
+        uint256 timestamp
+    );
+    event TokenRedirection(
+        address indexed destination,
+        address indexed token,
+        uint256 amount,
+        uint256 indexed blockNumber,
+        uint256 timestamp
+    );
```
Comment (shuklaayush) [inline contracts/MyStrategy.sol:87]: It's about time we check if block number and timestamp are needed to be logged in events or if they can be inferred while querying the events

File: contracts/MyStrategy.sol (hunk 2 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:532]: Good call. I guess if we ever wanted to avoid redirection we would have to remove the route path beforehand through governance, which is a cleaner and more transparent process. Will revert the change, thanks!

File: contracts/MyStrategy.sol (hunk 4 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:161]: That's a good question, will clarify with @Tritium-VLK if there could be multiple tokens for one recipient and, if this is the case, if the fees would be different per token.

File: contracts/MyStrategy.sol (hunk 5 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:87]: Good call, @petrovska-petro has been exploring the standardization of revenue tracking events, maybe he can give us more insights into this.

File: contracts/MyStrategy.sol (hunk 1 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:518]: This function will only be called if there was an amount claimed so `amount` will never be 0. The only reason the subtraction would lead to 0 is if the fee == MAX_BPS (100%). This shouldn't be the case but, since the fee can be set to this amount, it will indeed be good to protect against it - will fix.

File: contracts/MyStrategy.sol (hunk 5 above)
Comment (petrovska-petro) [reply contracts/MyStrategy.sol:87]: i am fan of adding them in the event itself since they make much easier the process of tracking those items while querying them. Even tho, we could say that if we leverage [eth_getLogs](https://www.quicknode.com/docs/ethereum/eth_getLogs) will provide the `block_number` and from it we could deduce the timestamp of that block, but will add up lots of operation to check the relation between block height -> timestamp, which each time will need to call [this](https://www.quicknode.com/docs/ethereum/eth_getBlockByNumber) into the rpc, overall making the process much slower to catch up all this information. so will advise to include them out of the box in the event itself to facilitate the querying of this events

File: contracts/MyStrategy.sol (hunk 2 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:532]: As a note, it just occurred to me that there may be a case in which a token is transferred to the strat by accident and this token has a redirection path set. In order to rescue this token we would have to atomically: remove path -> sweep -> add path back. Only issue with this is that it will be limited to governance only and not strategist but it is a very off scenario anyways.

Comment (shuklaayush) [reply contracts/MyStrategy.sol:532]: There's also a `sweepExtraToken` on the vault that might be useful for this

**Comment by sajanrajdev** [comment]
@shuklaayush @rayeaster, addressed all of your comments and adapted tests accordingly.

File: contracts/MyStrategy.sol (hunk 4 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol]: Mapped fees to the tokens as well, made more sense indeed.

File: contracts/MyStrategy.sol (hunk 3 above)
Comment (sajanrajdev) [reply contracts/MyStrategy.sol]: Ended up allowing for redirection using WETH as the token to map to.

**Review by rayeaster — APPROVED** [review]
The change looks good to me. Thanks

**Review by shuklaayush — COMMENTED** [review]
Two small comments. Otherwise looks good. Great job on the tests!

File: contracts/MyStrategy.sol [hunk 6]
```diff
@@ -347,17 +390,21 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
                 // ETH
                 uint256 difference = address(this).balance.sub(beforeBalance[i]);
                 if (difference > 0) {
+                    address recepient = bribesRedirectionPaths[address(WETH)];
```
[the remark this answers is not in the document]
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:393]: I had originally set it as such but then I realized that because the `_handleRewardTransfer` call at the following [line](https://github.com/Badger-Finance/vested-aura/blob/f8152243d0627e1945b0a6926e2564381caa7faf/contracts/MyStrategy.sol#L398) takes WETH as a parameter (and has to) and because now I am mapping the fee to the token (read from the token passed to this function) then the token mapped to and the token passed to the function had to be the same. Therefore, as long as the fee is mapped to the token (which makes sense), it has to be wETH for ETH. 

Good catch though. I think that as long as it is clear in the docs it should be fine. And again, I don't think we expect to redirect ETH anyways, this is just to have the option to do so.

File: contracts/MyStrategy.sol [hunk 7]
```diff
@@ -453,18 +500,43 @@ contract MyStrategy is BaseStrategy, ReentrancyGuardUpgradeable {
...
+        }
     }
 
     function _sweepRewardToken(address token) internal {
         _onlyNotProtectedTokens(token);
 
         uint256 toSend = IERC20Upgradeable(token).balanceOf(address(this));
-        _handleRewardTransfer(token, toSend);
+        address recepient = bribesRedirectionPaths[token];
+        _handleRewardTransfer(token, recepient, toSend);
```
[the remark this answers is not in the document]
Comment (sajanrajdev) [reply contracts/MyStrategy.sol:539]: This function doesn't notify the bribes processor on purpose (see [comments](https://github.com/Badger-Finance/vested-aura/blob/f8152243d0627e1945b0a6926e2564381caa7faf/contracts/MyStrategy.sol#L167)). The reason is that it can be called outside of a bribes round. For instance, on bveCVX we use it sweep the cvxFXS (received as an additional reward) into the processor to then handle it alongside with the bribes. We do it atomically with the claiming of bribes but we may want to do it separately at some point, if there's no bribes for whatever reason, and then rageQuit the processor to handle the cvxFXS manually. So not notifying will allow us to do so.

File: contracts/MyStrategy.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (shuklaayush) [reply contracts/MyStrategy.sol:393]: Ah missed that. Makes sense!

File: contracts/MyStrategy.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (shuklaayush) [reply contracts/MyStrategy.sol:539]: Makes sense!

**Review by shuklaayush — APPROVED** [review]
